compiler/src/iree/compiler/Codegen/SPIRV/SPIRVVectorize.cpp

```diff
@@ -404,6 +407,7 @@ class SPIRVVectorizePass : public SPIRVVectorizeBase<SPIRVVectorizePass> {
       vector::populateVectorMultiReductionLoweringPatterns(
           patterns, vector::VectorMultiReductionLowering::InnerParallel);
       vector::populateVectorTransposeLoweringPatterns(patterns, options);
+      vector::populateVectorGatherLoweringPatterns(patterns);
```

Comment:
Hmm, actually I think we want to do the `vector.gather` lowering after `vector.multi_reduction` lowering and before unrolling. The reason is that, the `vector.gather` lowering actually handles unrolling internally. It generates lots of extract/insert ops that should be cleaned up. We do such cleanups after unrolling. So if placed here it might be too late for certain cases. Could you move this upwards:

```c++
    // Similarly for vector.gather ops, whose lowering patterns unroll internally.
    {
      RewritePatternSet patterns(context);
      vector::populateVectorGatherLoweringPatterns(patterns);
      if (failed(applyPatternsAndFoldGreedily(funcOp, std::move(patterns)))) {
        return signalPassFailure();
      }
    }

    LLVM_DEBUG({
      llvm::dbgs() << "--- After lowering gather ops ---\n";
      funcOp.print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
      llvm::dbgs() << "\n\n";
    });
```

Reply:
Ah, I figured applying it after unrolling would let the gather have the same unrolled size as  nearby ops and then the extract/inserts from the lowering were inevitable. Let me try moving it.